compiler/rustc_lint/src/levels.rs

```diff
@@ -751,9 +751,26 @@ impl<'tcx> intravisit::Visitor<'tcx> for LintLevelMapBuilder<'tcx> {
     }
 
     fn visit_expr(&mut self, e: &'tcx hir::Expr<'tcx>) {
-        self.with_lint_attrs(e.hir_id, |builder| {
-            intravisit::walk_expr(builder, e);
-        })
+        match e.kind {
+            hir::ExprKind::Struct(qpath, fields, base_expr) => {
+                self.with_lint_attrs(e.hir_id, |builder| {
+                    builder.visit_qpath(qpath, e.hir_id, e.span);
+                    for field in fields {
```

Comment:
With all this custom visits, I wonder if we shouldn't add a `visit_expr_struct_field` to `hir::Visitor` and use it where appropriate.